processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java

```diff
@@ -88,21 +90,48 @@ public Filter toFilter()
       return new SelectorFilter(dimension, value);
     } else {
       final String valueOrNull = Strings.emptyToNull(value);
-      final Predicate<String> predicate = new Predicate<String>()
+      final Long valueAsLong = Longs.tryParse(value);
+
+      final DruidPredicateFactory predicateFactory = new DruidPredicateFactory()
       {
         @Override
-        public boolean apply(String input)
+        public Predicate<String> makeStringPredicate()
         {
-          return Objects.equals(valueOrNull, input);
+          return new Predicate<String>()
+          {
+            @Override
+            public boolean apply(String input)
+            {
+              return Objects.equals(valueOrNull, input);
+            }
+          };
         }
 
         @Override
-        public String toString()
+        public DruidLongPredicate makeLongPredicate()
         {
-          return value;
+          if (valueAsLong == null) {
+            return new DruidLongPredicate()
+            {
+              @Override
+              public boolean applyLong(long input)
+              {
+                return false;
+              }
+            };
+          } else {
+            return new DruidLongPredicate()
+            {
+              @Override
+              public boolean applyLong(long input)
+              {
+                return input == valueAsLong;
```

Comment:
@gianm From viewing the generated bytecode, the JVM will call Long.longValue() on every == comparison to unbox the Long; I've changed this to store the primitive value and compare against that directly to avoid the repeated .longValue() method calls